### DISCUSSION TITLE
RLS to view profiles of users that are in a team together

### DISCUSSION THREAD

**Discussion by 0x62** [supabase/supabase#11173] [Questions]
I have a `profiles` table, which contains every users email and name. Rows are created from triggers on `auth.users`.

```
CREATE TABLE IF NOT EXISTS "profiles" (
  "id" uuid PRIMARY KEY default uuid_generate_v4(),
  "provider" text NOT NULL default 'email',
  "name" text,
  "email" text NOT NULL,
  "avatarUrl" text,
  "createdAt" timestamptz NOT NULL default now()
);
```

Each user can have many projects, which they can invite other users to work on too.

```
CREATE TABLE IF NOT EXISTS "projects" (
  "id" uuid PRIMARY KEY default uuid_generate_v4(),
  ...
)

CREATE TABLE IF NOT EXISTS "projectInvites" (
  "id" uuid PRIMARY KEY default uuid_generate_v4(),
  "projectId" uuid NOT NULL references public.projects on delete cascade on update cascade,
  "userId" uuid NOT NULL references public.profiles on delete cascade on update cascade,
  "status" text NOT NULL default 'pending',
  "createdAt" timestamptz NOT NULL default now(),
  "actionedAt" timestamptz
);
```

I want to create an RLS policy that allows users to view the profile of other users they're in a project with, but I can't find many examples of this.

* Users should be able to view the `projects` table for all projects they've been invited to
* Users shouldn't be able to view any of the project contents (ie other tables with a `projectId` foreign key) unless the `projectInvites.status` is `accepted`
* Project members should be able to view the profiles of all invited users, whether accepted or not
* Once a user has accepted an invite, they should be able to view the profiles of all invited users

Based on other discussions, I've come up with the following so far:

```
create or replace function public.invited_to_project(user_id uuid, project_id uuid)
returns boolean
language plpgsql
security definer set search_path = public
as $$
begin
  RETURN (
    SELECT EXISTS (
      SELECT 1
      FROM "teamInvites"
      WHERE "userId" = user_id AND "projectId" = project_id
    )
  );
end;
$$;

create or replace function public.members_of_project(project_id uuid)
returns setof uuid
language sql
security definer set search_path = public
as $$
  SELECT "userId"
  FROM "teamInvites"
  WHERE "projectId" = project_id AND "status" = 'accepted'
$$;

create or replace function public.member_of_project(user_id uuid, project_id uuid)
returns boolean
language plpgsql
security definer set search_path = public
as $$
begin
  RETURN (
    SELECT user_id IN (SELECT members_of_project(project_id))
  );
end;
$$;

create policy "Users can view their project invites" on "projectInvites"
  for select using (
    is_staff()  OR invited_to_project(auth.uid(), "projectId")
  );

create policy "Users can view projects they have been invited to" on "projects"
  for select using (
    is_staff()  OR invited_to_project(auth.uid(), "projectId")
  );

--- pseudocode for a foreign key
create policy "Project members can view entities" on "projectEntity"
  for select using (
    is_staff()  OR member_of_project(auth.uid(), "projectId")
  );
```

What I'm stuck on is the RLS for profiles?

My starting point is something like this, but I'm not sure on the performance implications with lots of projects or if this is the right approach.

```
create or replace function public.in_project_with_user(my_id uuid, their_id uuid)
returns boolean
language plpgsql
security definer set search_path = public
as $$
begin
  RETURN (
    SELECT EXISTS (
      SELECT 1
      FROM "projects"
      WHERE member_of_project(my_id, id) AND invited_to_project(their_id, id)
    )
  );
end;

alter policy "Users can view their own profile" on profiles
  using (is_staff() OR auth.uid() = id OR in_project_with_user(auth.uid(), id));
```

**Comment by SebastianElustondo** [comment]
Since wrangler 4.27 there's a supported way to do exactly this: set `CLOUDFLARE_INCLUDE_PROCESS_ENV=true` and wrangler dev picks up the process environment as local vars, so your `env:` block just works:

```yaml
- run: npx -y wrangler dev --env dev
  env:
    CLOUDFLARE_INCLUDE_PROCESS_ENV: "true"
    G_KEY: ${{ secrets.G_KEY }}
    ADMIN_PK: ${{ secrets.ADMIN_PK }}
```

From 4.70 there's an even more explicit form: declare the secrets in the config and wrangler loads only those from `process.env` (or `.env` / `.dev.vars`), warning if one is missing:

```jsonc
// wrangler.jsonc
{ "secrets": { "required": ["G_KEY", "ADMIN_PK"] } }
```

`--var KEY:value` works too, but it puts the secret on the command line, and on a shared runner that's visible to anything that can list processes. Writing a `.dev.vars` in a step is fine on an ephemeral runner, but the env var approach needs no extra step.
